### PR TITLE
[Ready for review] Auto3D functional fixes in the BundleGen module

### PR DESCRIPTION
**Pull request by mingxin-zheng**
### Description
Part of the requested improvements #5085 

Planned fixes
- [x] fold number is not being updated in generate()
- [x] put absolute paths in the bundle configs
- [x] break the auto-generated auto_config into multiple files.

### Status
WIP

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/apps/auto3dseg/bundle_gen.py [hunk 1]
```diff
@@ -136,7 +142,7 @@ def fill_template_config(self, data_stats_filename, **kwargs):
         Args:
             data_stats_filename: filename of the data stats report (generated by DataAnalyzer)
         """
-        pass
+        raise NotImplementedError
```
Comment (Nic-Ma) [inline monai/apps/auto3dseg/bundle_gen.py:145]: Please add more error message, refer to:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/engines/workflow.py#L293

Thanks.

File: monai/apps/auto3dseg/bundle_gen.py [hunk 2]
```diff
@@ -153,21 +159,42 @@ def export_to_disk(self, output_path: str, algo_name: str, **kwargs):
...
+        logger.info(write_path)
+        self.output_path = write_path
+
+    def save_config_files(self, output_config_path, sections):
+        """
+        Save the auto-generated config files into multiple files.
+
+        Args:
+            output_config_path: path to save the files
+            subsections: the subsections that will be picked up and individually saved.
```
Comment (Nic-Ma) [inline monai/apps/auto3dseg/bundle_gen.py:182]: `subsections` does not match the function arg name `sections`.

Thanks.

File: monai/apps/auto3dseg/bundle_gen.py [hunk 3]
```diff
@@ -225,29 +259,51 @@ def get_score(self, *args, **kwargs):
...
         """
-        config_yaml = os.path.join(self.output_path, "configs", "algo_config.yaml")
+        config_yaml = os.path.join(self.output_path, "configs", "hyper_parameters.yaml")
         parser = ConfigParser()
         parser.read_config(config_yaml)
         ckpt_path = parser.get_parsed_content("ckpt_path", default=self.output_path)
 
         dict_file = ConfigParser.load_config_file(os.path.join(ckpt_path, "progress.yaml"))
-        return dict_file["best_avg_dice_score"]  # TODO: define the format of best scores' list and progress.yaml
+        return dict_file[-1]["best_avg_dice_score"]  # TODO: define the format of best scores' list and progress.yaml
```
Comment (Nic-Ma) [inline monai/apps/auto3dseg/bundle_gen.py:268]: Please add 1 comment to explain the `-1` index here.

Thanks.

File: monai/apps/auto3dseg/bundle_gen.py [hunk 4]
```diff
@@ -153,21 +159,42 @@ def export_to_disk(self, output_path: str, algo_name: str, **kwargs):
...
         os.makedirs(output_config_path, exist_ok=True)
-        output_config_file = os.path.join(output_config_path, "algo_config.yaml")
-        ConfigParser.export_config_file(self.cfg.config, output_config_file, fmt="yaml", default_flow_style=None)
+
+        config_body = self.cfg.get()
+        for s in sections:
+            config_section_file = os.path.join(output_config_path, s + ".yaml")
+            config_section = _pop_args(config_body, s)
+            section_body = {}
+            section_body[s] = config_section[0]
```
Comment (Nic-Ma) [inline monai/apps/auto3dseg/bundle_gen.py:192]: Please add 1 comment to explain the `0` index here.

Thanks.

**Review by Nic-Ma — APPROVED** [review]
Thanks for the quick update according to test feedback.
Overall looks good to me, put minor comments inline.

Thanks.

File: monai/apps/auto3dseg/bundle_gen.py (hunk 1 above)
Comment (mingxin-zheng) [reply monai/apps/auto3dseg/bundle_gen.py:145]: Fixed. Thanks!

File: monai/apps/auto3dseg/bundle_gen.py [hunk 5]
```diff
@@ -136,7 +142,7 @@ def fill_template_config(self, data_stats_filename, **kwargs):
         Args:
             data_stats_filename: filename of the data stats report (generated by DataAnalyzer)
         """
-        pass
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
```
Comment (wyli) [inline monai/apps/auto3dseg/bundle_gen.py:145]: I think it's nicer to use `pass` to make this method call optional, the bundle template might be ready to train without this template filling..

Comment (mingxin-zheng) [reply monai/apps/auto3dseg/bundle_gen.py:145]: Yes that's possible. Filling the template is optional. The user can simply write a train.py without any configs.

Comment (mingxin-zheng) [reply monai/apps/auto3dseg/bundle_gen.py:145]: Added more doc-strings to explain this behavior

**Pull request merged**